### ISSUE TITLE
strictNullChecks false-positive when returning or breaking from within loop

### ISSUE THREAD

**Issue by sccolbert** [microsoft/TypeScript#14143]
Compile the following with `--strictNullChecks`

```ts
  function findScrollTarget(event: MouseEvent): { element: Element } | null {
    let x = event.clientX;
    let y = event.clientY;
    let element: Element | null = document.elementFromPoint(x, y);
    for (; element; element = element.parentElement) {
      return { element };
    }
    return null;
  }
```

**Expected behavior:**
This should compile, but doesn't with the following error: 
`src/index.ts(796,31): error TS2531: Object is possibly 'null'.`

That line number and position corresponds to to the `element.parentElement` expression in the third clause of the for-loop in my example.

This example also fails with the same error:

```ts
  function findScrollTarget(event: MouseEvent): { element: Element } | null {
    let x = event.clientX;
    let y = event.clientY;
    let result: { element: Element } | null = null;
    let element: Element | null = document.elementFromPoint(x, y);
    for (; element; element = element.parentElement) {
      result = { element };
      break;
    }
    return result;
  }
```
 
However, this example compiles fine, which doesn't escape early from the loop.

```ts
  function findScrollTarget(event: MouseEvent): { element: Element } | null {
    let x = event.clientX;
    let y = event.clientY;
    let result: { element: Element } | null = null;
    let element: Element | null = document.elementFromPoint(x, y);
    for (; element; element = element.parentElement) {
      result = { element };
    }
    return result;
  }
```

**Comment by RyanCavanaugh** [comment]
Seems like we ought to flag the update clauses as unreachable code instead?

**Comment by RyanCavanaugh** [comment]
Duplicate #26914
